unix_integration/src/resolver.rs

```diff
@@ -95,21 +100,71 @@ where
         gid_attr_map: UidAttr,
         allow_id_overrides: Vec<String>,
     ) -> Result<Self, ()> {
+        let hsm = Mutex::new(hsm);
+        let mut hsm_lock = hsm.lock().await;
+
         // setup and do a migrate.
-        {
-            let dbtxn = db.write().await;
-            dbtxn.migrate().map_err(|_| ())?;
-            dbtxn.commit().map_err(|_| ())?;
-        }
+        let dbtxn = db.write().await;
+        dbtxn.migrate().map_err(|_| ())?;
+        dbtxn.commit().map_err(|_| ())?;
+
+        // Setup our internal keys
+        let dbtxn = db.write().await;
+
+        let loadable_hmac_key = match dbtxn.get_hsm_hmac_key() {
+            Ok(Some(hmk)) => hmk,
+            Ok(None) => {
+                // generate a new key.
+                let loadable_hmac_key = hsm_lock.hmac_key_create(&machine_key).map_err(|err| {
+                    error!(?err, "Unable to create hmac key");
+                })?;
+
+                dbtxn
+                    .insert_hsm_hmac_key(&loadable_hmac_key)
+                    .map_err(|err| {
+                        error!(?err, "Unable to persist hmac key");
+                    })?;
+
+                loadable_hmac_key
+            }
+            Err(err) => {
+                error!(?err, "Unable to retrieve loadable hmac key from db");
+                return Err(());
+            }
+        };
+
+        let hmac_key = hsm_lock
+            .hmac_key_load(&machine_key, &loadable_hmac_key)
+            .map_err(|err| {
+                error!(?err, "Unable to load hmac key");
+            })?;
+
+        // Ask the client what keys it wants the HSM to configure.
+        // make a key store
+        let mut ks = KeyStore::new(&dbtxn);
+
+        client
+            .configure_hsm_keys(&mut ks, &mut **hsm_lock.deref_mut(), &machine_key)
+            .await
+            .map_err(|err| {
+                error!(?err, "Client was unable to configure hsm keys");
+            })?;
+
+        drop(hsm_lock);
+
+        dbtxn.commit().map_err(|_| ())?;
 
         if pam_allow_groups.is_empty() {
-            eprintln!("Will not be able to authorise user logins, pam_allow_groups config is not configured.");
+            warn!("Will not be able to authorise user logins, pam_allow_groups config is not configured.");
```

Comment:
This really feels like it should just make the program bail 🤔

Reply:
Yeah, I was thinking that too. But I'm not sure either. Is there a reason you'd install a kanidm resolver and *not* need pam_allow_groups to have entries? My only example here is I have a samba server in a container and it needs to *resolve* users, but not authenticate them, so this is empty in it's config iirc.

Reply:
Ah, I guess that'd work. Feels kinda icky to be using the IDM for one thing and not another 🤣

Reply:
It's still IDM though, it's resolving identities. So kinda a bit in scope :)